README.md
[![pub package](https://img.shields.io/pub/v/flutter_candlesticks.svg)](https://pub.dartlang.org/packages/flutter_candlesticks)

# flutter_candlesticks

Elegant OHLC Candlestick and Trade Volume charts for Flutter

## Usage

Install for Flutter [with pub](https://pub.dartlang.org/packages/flutter_candlesticks#-installing-tab-).

| Property           | Description                                                               |
|--------------------|---------------------------------------------------------------------------|
| data               | Required. List of maps containing open, high, low, close and volumeto     |
| enableGridLines    | Required. Enable or disable grid lines                                    |
| volumeProp         | Required. Proportion of container to be given to volume bars              |
| candleSpacing      | Default `1.0`. Width of spacing between candles                           |
| gridLineAmount     | Default `5`. Number of grid lines to draw. Labels automatically assigned  |
| gridLineWidth      | Default `0.5`. Width of grid lines                                        |
| gridLineColor      | Default `Colors.grey`. Color of grid lines                                |
| gridLineLabelColor | Default `Colors.grey`. Color of grid line labels                          |
| labelPrefix        | Default `"$"`. Prefix before grid line labels.                            |
| increasePaint      | Default `green stroke`. Paint of increasing candles.                      |
| decreasePaint      | Default `red fill`. Paint of decreasing candles.                          |
| volumePaint        | Default `corresponding candlestick color`. Paint of volume bars           |

## Examples

### No Grid Lines

```dart
new OHLCVGraph(
    data: sampleData,
    enableGridLines: false,
    volumeProp: 0.2
)
```

<img src="screenshots/white_large.png" width="50%"><img src="screenshots/dark_large.png" width="50%">

> Candle size dynamically changes by amount of data

<img src="screenshots/white_small.png" width="50%"><img src="screenshots/dark_small.png" width="50%">


### Grid Lines

```dart
new OHLCVGraph(
    data: sampleData,
    enableGridLines: true,
    volumeProp: 0.2,
    gridLineAmount: 5,
    gridLineColor: Colors.grey[300],
    gridLineLabelColor: Colors.grey
)
```

<img src="screenshots/white_small_gridlines.png" width="50%"><img src="screenshots/dark_large_gridlines.png" width="50%">


### Style Example

```dart
new OHLCVGraph(
    data: sampleData,
    enableGridLines: true,
    volumeProp: 0.2,
    gridLineAmount: 5,
    gridLineColor: Colors.grey[300],
    gridLineLabelColor: Colors.grey
    decreasePaint: new Paint()
      ..color = Colors.red
      ..strokeWidth = 1.0
      ..style = PaintingStyle.fill
    ,
    increasePaint: new Paint()
      ..color = Colors.green
      ..strokeWidth = 1.0
      ..style = PaintingStyle.fill
    ,
    volumePaint: new Paint()
      ..color = Colors.grey
      ..style = PaintingStyle.fill,
    candleSpacing: 1.2,
)
```

<img src="screenshots/custom_colors.png" width="50%"><img src="screenshots/custom_colors_dark.png" width="50%">


### Full App Example
```dart
import 'package:flutter/material.dart';
import 'package:flutter_candlesticks/flutter_candlesticks.dart';

void main() {
  List sampleData = [
    {"open":50.0, "high":100.0, "low":40.0, "close":80, "volumeto":5000.0},
    {"open":80.0, "high":90.0, "low":55.0, "close":65, "volumeto":4000.0},
    {"open":65.0, "high":120.0, "low":60.0, "close":90, "volumeto":7000.0},
    {"open":90.0, "high":95.0, "low":85.0, "close":80, "volumeto":2000.0},
    {"open":80.0, "high":85.0, "low":40.0, "close":50, "volumeto":3000.0},
  ];

  runApp(
    new MaterialApp(
      home: new Scaffold(
        body: new Center(
          child: new Container(
            height: 500.0,
            child: new OHLCVGraph(
                data: sampleData,
                enableGridLines: false,
                volumeProp: 0.2
            ),
          ),
        ),
      )
    )
  );
}
```


CHANGELOG.md
## [0.1.4] - 2019-04-28

* Fix unused label prefix

## [0.1.3] - 2019-04-28

* Dart 2

## [0.1.2] - 2018-08-31

* Update version constraints for Dart 2.1.0

## [0.1.1] - 2018-08-06

* Added increaseColor and decreaseColor options

## [0.1.0] - 2018-07-09

* Fix gridLineWidth not being used
* Better grid line label text handling

## [0.0.9] - 2018-04-26

* Fix label value bug
* Added labelPrefix param

## [0.0.8] - 2018-04-26

* Update to Flutter SDK 0.3.1

## [0.0.7] - 2018-04-20

* Moved labels to right of grid lines

## [0.0.6] - 2018-04-12

* Improved labeling efficiency

## [0.0.5] - 2018-04-10

* Corrected README.md

## [0.0.4] - 2018-04-10

* Fix screenshots for pub

## [0.0.3] - 2018-04-10

* Update README.md
* Screenshots
* Fix volume label color

## [0.0.2] - 2018-04-10

* LICENSE
* Example

## [0.0.1] - 2018-04-09

* Initial Release


lib/flutter_candlesticks.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

class OHLCVGraph extends StatelessWidget {
  OHLCVGraph({
    Key key,
    @required this.data,
    this.candleSpacing = 1.0,
    this.fallbackHeight = 100.0,
    this.fallbackWidth = 300.0,
    this.gridLineColor = Colors.grey,
    this.gridLineAmount = 5,
    this.gridLineWidth = 0.5,
    this.gridLineLabelColor = Colors.grey,
    this.labelPrefix = "\$",
    this.decreasePaint,
    this.increasePaint,
    this.volumePaint,
    @required this.enableGridLines,
    @required this.volumeProp,
  })  : assert(data != null),
        assert(candleSpacing != null),
        super(key: key);

  /// OHLCV data to graph
  /// List of Maps containing open, high, low, close and volumeto
  /// Example: [["open" : 40.0, "high" : 75.0, "low" : 25.0, "close" : 50.0, "volumeto" : 5000.0}, {...}]
  final List data;

  /// spacing between lines
  final double candleSpacing;

  /// Enable or disable grid lines
  final bool enableGridLines;

  /// Color of grid lines and label text
  final Color gridLineColor;
  final Color gridLineLabelColor;

  /// Number of grid lines
  final int gridLineAmount;

  /// Width of grid lines
  final double gridLineWidth;

  /// Proportion of paint to be given to volume bar graph
  final double volumeProp;

  /// If graph is given unbounded space,
  /// it will default to given fallback height and width
  final double fallbackHeight;
  final double fallbackWidth;

  /// Symbol prefix for grid line labels
  final String labelPrefix;

  /// Increasing candle paint
  final Paint increasePaint;

  /// Decreasing candle paint
  final Paint decreasePaint;

  /// Volume candle paint (defaults to Increasing/Decreasing paints if empty)
  final Paint volumePaint;

  @override
  Widget build(BuildContext context) {
    return new LimitedBox(
      maxHeight: fallbackHeight,
      maxWidth: fallbackWidth,
      child: new CustomPaint(
        size: Size.infinite,
        painter: new _OHLCVPainter(
          data,
          candleSpacing: candleSpacing,
          gridLineColor: gridLineColor,
          gridLineAmount: gridLineAmount,
          gridLineWidth: gridLineWidth,
          gridLineLabelColor: gridLineLabelColor,
          enableGridLines: enableGridLines,
          volumeProp: volumeProp,
          labelPrefix: labelPrefix,
          decreasePaint: decreasePaint ?? (new Paint()
            ..color = Colors.red
            ..strokeWidth = 1.0
            ..style = PaintingStyle.fill
          ),
          increasePaint: increasePaint ?? (new Paint()
            ..color = Colors.green
            ..strokeWidth = 1.0
            ..style = PaintingStyle.stroke
          ),
          volumePaint: volumePaint
        ),
      ),
    );
  }
}

class _OHLCVPainter extends CustomPainter {
  _OHLCVPainter(
    this.data, {
    @required this.candleSpacing,
    @required this.enableGridLines,
    @required this.gridLineColor,
    @required this.gridLineAmount,
    @required this.gridLineWidth,
    @required this.gridLineLabelColor,
    @required this.volumeProp,
    @required this.labelPrefix,
    @required this.decreasePaint,
    @required this.increasePaint,
    @required this.volumePaint,
  });

  final List data;
  final double candleSpacing;
  final bool enableGridLines;
  final Color gridLineColor;
  final int gridLineAmount;
  final double gridLineWidth;
  final Color gridLineLabelColor;
  final String labelPrefix;

  final double volumeProp;

  final Paint decreasePaint;
  final Paint increasePaint;
  final Paint volumePaint;

  double _min;
  double _max;
  double _maxVolume;

  List<TextPainter> gridLineTextPainters = [];
  TextPainter maxVolumePainter;

  numCommaParse(number) {
    return number.round().toString().replaceAllMapped(
        new RegExp(r'(\d{1,3})(?=(\d{3})+(?!\d))'), (Match m) => "${m[1]},");
  }

  update() {
    _min = double.infinity;
    _max = -double.infinity;
    _maxVolume = -double.infinity;
    for (var i in data) {
      if (i["high"] > _max) {
        _max = i["high"].toDouble();
      }
      if (i["low"] < _min) {
        _min = i["low"].toDouble();
      }
      if (i["volumeto"] > _maxVolume) {
        _maxVolume = i["volumeto"].toDouble();
      }
    }

    if (enableGridLines) {
      double gridLineValue;
      for (int i = 0; i < gridLineAmount; i++) {
        // Label grid lines
        gridLineValue = _max - (((_max - _min) / (gridLineAmount - 1)) * i);

        String gridLineText;
        if (gridLineValue < 1) {
          gridLineText = gridLineValue.toStringAsPrecision(4);
        } else if (gridLineValue < 999) {
          gridLineText = gridLineValue.toStringAsFixed(2);
        } else {
          gridLineText = gridLineValue.round().toString().replaceAllMapped(
              new RegExp(r'(\d{1,3})(?=(\d{3})+(?!\d))'),
              (Match m) => "${m[1]},");
        }

        gridLineTextPainters.add(new TextPainter(
            text: new TextSpan(
                text: labelPrefix + gridLineText,
                style: new TextStyle(
                    color: gridLineLabelColor,
                    fontSize: 10.0,
                    fontWeight: FontWeight.bold)),
            textDirection: TextDirection.ltr));
        gridLineTextPainters[i].layout();
      }

      // Label volume line
      maxVolumePainter = new TextPainter(
          text: new TextSpan(
              text: "\$" + numCommaParse(_maxVolume),
              style: new TextStyle(
                  color: gridLineLabelColor,
                  fontSize: 10.0,
                  fontWeight: FontWeight.bold)),
          textDirection: TextDirection.ltr);
      maxVolumePainter.layout();
    }
  }

  @override
  void paint(Canvas canvas, Size size) {
    if (_min == null || _max == null || _maxVolume == null) {
      update();
    }

    final double volumeHeight = size.height * volumeProp;
    final double volumeNormalizer = volumeHeight / _maxVolume;

    double width = size.width;
    final double height = size.height * (1 - volumeProp);

    double gridLineY;

    if (enableGridLines) {
      width = size.width - gridLineTextPainters[0].text.text.length * 6;
      Paint gridPaint = new Paint()
        ..color = gridLineColor
        ..strokeWidth = gridLineWidth;

      double gridLineDist = height / (gridLineAmount - 1);

      // Draw grid lines
      for (int i = 0; i < gridLineAmount; i++) {
        gridLineY = (gridLineDist * i).round().toDouble();
        canvas.drawLine(new Offset(0.0, gridLineY),
            new Offset(width, gridLineY), gridPaint);

        // Label grid lines
        gridLineTextPainters[i]
            .paint(canvas, new Offset(width + 2.0, gridLineY - 6.0));
      }
    }

    final double heightNormalizer = height / (_max - _min);
    final double rectWidth = width / data.length;

    double rectLeft;
    double rectTop;
    double rectRight;
    double rectBottom;

    Paint rectPaint;

    // Loop through all data
    for (int i = 0; i < data.length; i++) {
      rectLeft = (i * rectWidth);
      rectRight = ((i + 1) * rectWidth);

      double volBarTop = (height + volumeHeight) -
          (data[i]["volumeto"] * volumeNormalizer);
      double volBarBottom = height + volumeHeight;

      bool isDecreasing = (data[i]["open"] > data[i]["close"]);

      rectPaint = isDecreasing ? decreasePaint : increasePaint;

      rectTop = height - (data[i][isDecreasing ? "open" : "close"] - _min)
          * heightNormalizer;
      rectBottom = height - (data[i][isDecreasing ? "close" : "open"] - _min)
          * heightNormalizer;

      // Draw candlestick body
      var ocRect = new Rect.fromLTRB(
          rectLeft + candleSpacing, rectTop, rectRight - candleSpacing, rectBottom);
      if (rectPaint.style == PaintingStyle.stroke)
        ocRect = ocRect.deflate(rectPaint.strokeWidth/2);
      canvas.drawRect(ocRect, rectPaint);

      // Draw volume bars
      Rect volumeRect = new Rect.fromLTRB(
          rectLeft + candleSpacing, volBarTop, rectRight - candleSpacing, volBarBottom);
      if (volumePaint != null && volumePaint.style == PaintingStyle.stroke)
        volumeRect = volumeRect.deflate(volumePaint.strokeWidth/2);
      canvas.drawRect(volumeRect, volumePaint ?? rectPaint);

      // Draw low/high candlestick wicks
      double low = height - (data[i]["low"] - _min) * heightNormalizer;
      double high = height - (data[i]["high"] - _min) * heightNormalizer;
      canvas.drawLine(
          new Offset(rectLeft + (rectWidth) / 2, rectBottom),
          new Offset(rectLeft + (rectWidth) / 2, low),
          rectPaint);
      canvas.drawLine(
          new Offset(rectLeft + (rectWidth) / 2, rectTop),
          new Offset(rectLeft + (rectWidth) / 2, high),
          rectPaint);
    }

    if (enableGridLines) {
      // Label volume line
      maxVolumePainter.paint(canvas, new Offset(0.0, gridLineY + 2.0));
    }
  }

  @override
  bool shouldRepaint(_OHLCVPainter old) {
    return data != old.data ||
        candleSpacing != old.candleSpacing ||
        enableGridLines != old.enableGridLines ||
        gridLineColor != old.gridLineColor ||
        gridLineAmount != old.gridLineAmount ||
        gridLineWidth != old.gridLineWidth ||
        volumeProp != old.volumeProp ||
        gridLineLabelColor != old.gridLineLabelColor;
  }
}


example/example.dart
import 'package:flutter/material.dart';
import 'package:flutter_candlesticks/flutter_candlesticks.dart';

void main() {
  List sampleData = [
    {"open":50.0, "high":100.0, "low":40.0, "close":80, "volumeto":5000.0},
    {"open":80.0, "high":90.0, "low":55.0, "close":65, "volumeto":4000.0},
    {"open":65.0, "high":120.0, "low":60.0, "close":90, "volumeto":7000.0},
    {"open":90.0, "high":95.0, "low":85.0, "close":80, "volumeto":2000.0},
    {"open":80.0, "high":85.0, "low":40.0, "close":50, "volumeto":3000.0},

    {"open":50.0, "high":100.0, "low":40.0, "close":80, "volumeto":5000.0},
    {"open":80.0, "high":90.0, "low":55.0, "close":65, "volumeto":4000.0},
    {"open":65.0, "high":120.0, "low":60.0, "close":90, "volumeto":7000.0},
    {"open":90.0, "high":95.0, "low":85.0, "close":80, "volumeto":2000.0},
    {"open":80.0, "high":85.0, "low":40.0, "close":50, "volumeto":3000.0},

    {"open":50.0, "high":100.0, "low":40.0, "close":80, "volumeto":5000.0},
    {"open":80.0, "high":90.0, "low":55.0, "close":65, "volumeto":4000.0},
    {"open":65.0, "high":120.0, "low":60.0, "close":90, "volumeto":7000.0},
    {"open":90.0, "high":95.0, "low":85.0, "close":80, "volumeto":2000.0},
    {"open":80.0, "high":85.0, "low":40.0, "close":50, "volumeto":3000.0},
  ];

  runApp(
    new MaterialApp(
      home: new Scaffold(
        body: new Center(
          child: new Container(
            height: 500.0,
            child: new OHLCVGraph(
                data: sampleData,
                enableGridLines: false,
                volumeProp: 0.2
            ),
          ),
        ),
      )
    )
  );
}